src/protocol/http/org/apache/jmeter/protocol/http/control/CacheManager.java

```diff
@@ -127,19 +163,28 @@ public Date getExpires() {
      * @param res result
      */
     public void saveDetails(URLConnection conn, HTTPSampleResult res){
-        if (isCacheable(res) && !hasVaryHeader(conn)){
+        if (isCacheable(res)){
             String lastModified = conn.getHeaderField(HTTPConstants.LAST_MODIFIED);
             String expires = conn.getHeaderField(HTTPConstants.EXPIRES);
             String etag = conn.getHeaderField(HTTPConstants.ETAG);
             String url = conn.getURL().toString();
             String cacheControl = conn.getHeaderField(HTTPConstants.CACHE_CONTROL);
             String date = conn.getHeaderField(HTTPConstants.DATE);
-            setCache(lastModified, cacheControl, expires, etag, url, date);
+            setCache(lastModified, cacheControl, expires, etag, url, date, getVaryHeader(conn.getHeaderField(HTTPConstants.VARY), res.getRequestHeaders()));
         }
     }
 
-    private boolean hasVaryHeader(URLConnection conn) {
-        return conn.getHeaderField(HTTPConstants.VARY) != null;
+    private Pair<String, String> getVaryHeader(String headerName, String reqHeaders) {
+        if (headerName == null) {
+            return null;
+        }
+        for (String headerLine: reqHeaders.split("\n")) {
+            if (headerLine.startsWith(headerName + ": ")) {
```

Comment:
I am thinking of splitting the `headerLine`, so that I can easier compare the `headerName` ignoring the case.

And as undera pointed out, a vary header can have multiple values (separated by comma), so this logic will have to be changed anyway.